packages/teleterm/src/services/grpcCerts/generateGrpcCerts.ts

```diff
@@ -0,0 +1,39 @@
+import fs from 'fs/promises';
+import path from 'path';
+import mkcert from 'mkcert';
+
+export async function generateGrpcCerts(certsDir: string): Promise<void> {
+  function getPath(fileName: string): string {
+    return path.join(certsDir, fileName);
+  }
+
+  const ca = await mkcert.createCA({
+    organization: 'localhost',
+    countryCode: 'US',
+    state: 'CA',
+    locality: 'Oakland',
+    validityDays: 365,
+  });
+
+  const server = await mkcert.createCert({
+    domains: ['localhost'],
+    validityDays: 365,
+    caKey: ca.key,
+    caCert: ca.cert,
+  });
+
+  const client = await mkcert.createCert({
+    domains: ['localhost'],
+    validityDays: 365,
+    caKey: ca.key,
+    caCert: ca.cert,
+  });
+
+  await Promise.all([
+    fs.writeFile(getPath('ca.crt'), ca.cert),
+    fs.writeFile(getPath('server.crt'), server.cert),
+    fs.writeFile(getPath('server.key'), server.key),
+    fs.writeFile(getPath('client.crt'), client.cert),
+    fs.writeFile(getPath('client.key'), client.key),
```

Comment:
My idea was to avoid ever having private key material on disk and to essentially tie the authorization to the ability to write in what is (hopefully) a private directory (to overwrite the certificate) rather than just the ability to read in that directory.

This warrants some testing, but I'm pretty sure that if you have a certificate in the trust store and the end-entity cert presented by the other end of a TLS connection is _that_ certificate then any authentication is assumed to pass, so you can just put a self-signed certificate in `rootCerts`/`ClientCAs` without having to generate a certificate tree.

If you don't want tsh to generate its own cert/key then you should be generating a single self-signed cert/key, store it on disk, and use it for both ends of the connection.

Reply:
Cool, I wasn't aware that I do not have to create that whole certificate tree!
Thanks to this, I can simplify the `makeCerts.ts` module.

>If you don't want tsh to generate its own cert/key then you should be generating a single self-signed cert/key, store it on disk, and use it for both ends of the connection.

Still, it seems to me that the simplest solution is to have one place in the app that generates certificates.

Reply:
How does `tshd` get told about this "CA" to verify client connections? Is there any chance that it will also trust it for outbound connections?

If that's not a possibility then LGTM, although the cert generation could be slimmed down a fair bit - we shouldn't need any SANs, or even _any_ field in the subject (although a CN of `teleport-connect-internal` or something along those lines could be useful if it ends up being printed anywhere).

Basically the equivalent of
```
% openssl req -new -x509 -newkey ec:<( openssl ecparam -name prime256v1 ) -keyout key.pem -nodes -out cert.pem -subj /CN=teleport-connect -set_serial 0 -days 3650
Generating a 256 bit EC private key
writing new private key to 'key.pem'
-----
espadolini@cantor:~% openssl x509 -noout -text -in cert.pem
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number: 0 (0x0)
    Signature Algorithm: ecdsa-with-SHA256
        Issuer: CN=teleport-connect
        Validity
            Not Before: Jul 19 09:44:09 2022 GMT
            Not After : Jul 16 09:44:09 2032 GMT
        Subject: CN=teleport-connect
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:10:8e:b7:f4:42:d6:c9:c5:df:dc:94:42:ea:12:
                    8b:cd:20:e1:0f:1c:91:c0:95:87:41:84:55:42:32:
                    b2:36:dc:78:0d:9c:21:75:82:0b:fe:7a:85:cc:c5:
                    e6:91:9a:29:ef:14:68:99:88:07:91:49:87:07:eb:
                    9d:bc:70:78:66
                ASN1 OID: prime256v1
                NIST CURVE: P-256
    Signature Algorithm: ecdsa-with-SHA256
         30:45:02:20:1a:13:1d:ea:50:ab:dd:02:04:98:9e:86:0c:7a:
         ff:86:db:ed:19:65:3f:e9:3f:f7:76:37:a2:26:d7:3b:bf:06:
         02:21:00:81:be:0a:eb:f2:c7:50:09:6f:a1:c1:af:8e:98:93:
         df:1f:e4:1d:08:b9:04:e3:a4:16:d7:ff:9b:2b:93:72:71
```

Reply:
Thx, I removed unneeded fields in the cert.

>How does tshd get told about this "CA" to verify client connections?
[`tshd` reads `certs.crt` from the `certs` directory](https://github.com/gravitational/teleport/blob/d959e315cd2e865d68ab3a27659e9b0b01028535/lib/teleterm/apiserver/apiserver.go#L121) which is located in user space. Only the user should be able to create a connection. 

>Is there any chance that it will also trust it for outbound connections?
No, it should only trust `localhost` connections that come from a user who has access to the certs.

I tested using `teleport-connect-internal` as CN, but it ended with the error:
`Hostname/IP does not match certificate's altnames: Host: localhost. is not cert's CN: teleport-connect-internal`

Reply:
I thought more about it and decided to not save the private key on a disk to increase the security and enable TLS only for TCP connections to save the resources.
Thanks @espadolini for pointing this out!

Now we store these three files:
- `shared_server.crt` generated by shared process 
- `tsh_server.crt` generated by tsh process  
- `client.crt` generated by window preload script

Private keys are stored in the memory of each process.

Then the servers wait for the client's cert, and the client waits for the servers' cert. Creating a `tsh` server is a bit simpler because we can use the [`GetConfigForClient` function, which reads the client's cert from disk when it tries to establish a connection](https://github.com/gravitational/teleport/blob/a15c606bf0ed568fdf6a9e6e59d6e5ce93f21d16/lib/teleterm/apiserver/apiserver.go#L150). In the case of a shared server, this is no that option, we have to wait with the creation of the server until the client's cert is read.

I encapsulated the logic involved in generating & reading certs, and then creating `ServerCredentials` & `ClientCredentials` in the `grpcCredentials` module. This makes the shared server and preload unaware of all the machinery needed to create the credentials. In addition, I could easily disable the use of TLS for UDS connections where it is not needed. 

Waiting for files is done using node.js `watch`, this should be the most efficient option. We can't watch a file that doesn't exist, so instead the entire directory is watched. Then we wait for the first event with a non-empty file with the specified name. After that, the watcher is closed.